includes.php

```diff
@@ -66,6 +66,191 @@
 		</header>
 		<main>';
 
+$mysqli = new mysqli( 'localhost', 'patchdemo', 'patchdemo', 'patchdemo' );
+if ( $mysqli->connect_error ) {
+	die( $mysqli->connect_error );
+}
+
+function insert_wiki_data( string $wiki, string $creator, int $created, array $patches = [] ) {
+	global $mysqli;
+	$stmt = $mysqli->prepare( '
+		INSERT INTO wikis
+		(wiki, creator, created, patches)
+		VALUES(?, ?, NOW(), ?)
+		ON DUPLICATE KEY UPDATE
+		patches = ?
+	' );
+	if ( !$stmt ) {
+		echo $mysqli->error;
+	}
+	$patches = json_encode( $patches );
+	$stmt->bind_param( 'ssss', $wiki, $creator, $patches, $patches );
+	$stmt->execute();
+	$stmt->close();
+}
+
+function wiki_add_announced_tasks( string $wiki, array $announcedTasks ) {
+	global $mysqli;
+	$stmt = $mysqli->prepare( 'UPDATE wikis SET announcedTasks = ? WHERE wiki = ?' );
+	$announcedTasks = json_encode( $announcedTasks );
+	$stmt->bind_param( 'ss', $accouncedTasks, $wiki );
+	$stmt->execute();
+	$stmt->close();
+}
+
+function get_wiki_data( string $wiki ) {
+	global $mysqli;
+
+	$stmt = $mysqli->prepare( '
+		SELECT wiki, creator, UNIX_TIMESTAMP( created ) created, patches, announcedTasks
+		FROM wikis WHERE wiki = ?
+	' );
+	if ( !$stmt ) {
+		echo $mysqli->error;
+	}
+	$stmt->bind_param( 's', $wiki );
+	$stmt->execute();
+	$res = $stmt->get_result();
+	$data = $res->fetch_assoc();
+	$stmt->close();
+
+	if ( !$data ) {
+		throw new Error( 'Wiki not found: ' . $wiki );
+	}
+
+	// Decode JSON
+	$data['patches'] = json_decode( $data['patches'] ) ?: [];
+	$data['announcedTasks'] = json_decode( $data['announcedTasks'] ) ?: [];
+
+	// Populate patch list
+	$patchList = [];
+	$linkedTasks = [];
+	if ( $data['patches'] ) {
+		foreach ( $data['patches'] as $patch ) {
+			[ $r, $p ] = explode( ',', $patch );
+			$patchData = get_patch_data( $r, $p );
+			$patchList[$patch] = get_patch_data( $r, $p );
```

Comment:
Calling `get_patch_data()` twice will do SQL and Gerrit API queries twice?

Reply:
Looks like a typo. Fixed.